File: src/main/java/alfheim/api/block/tile/SubTileEntity.java
package alfheim.api.block.tile;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import alexsocol.asjlib.ASJUtilities;
import alexsocol.asjlib.math.Vector3;
import alfheim.api.AlfheimAPI;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.AxisAlignedBB;
import net.minecraft.util.MathHelper;
import net.minecraft.world.World;

// Used for anomalies
public abstract class SubTileEntity {

	/** The Tag items should use to store which sub tile they are. **/
	public static final String TAG_TYPE = "type";
	public static final String TAG_TICKS = "ticks";
	public static final ArrayList<Object> EMPTY_LIST = new ArrayList<Object>(0);
	public Random rand = new Random();
	
	public int ticks;
	
	public TileEntity superTile;
	
	/** optional update method for particles or other stuff */
	protected void update() {}
	
	public final void updateEntity(List<Object> l) {
		update();
		
		effect: {
			if (l == null || l.isEmpty()) break effect;
			while(l.contains(null)) l.remove(null);
			if (l.isEmpty()) break effect;
		
			for (Object target : l) performEffect(target);
		}
		
		ticks++;
		//ASJUtilities.chatLog("Done");
	}
	
	public abstract List<Object> getTargets();
	
	public abstract void performEffect(Object target);
	
	public static final int NONE	= 0,			//	0b00000 - fully compatible, do not use this unless you know what you are doing 
							MOTION	= 1,			//	0b00001 - motion manipulation	- gravity 
							HEALTH	= 2,			//	0b00010 - health manipulation	- damaging
							MANA	= 4,			//	0b00100 - mana manipulation		- drain mana
							TIME	= 8,			//	0b01000 - ticks manipulation	- time speedup
							SPACE	= 17,			//	0b10001 - space manipulation	- teleportation		- also incompatible with motion
							ALL		= 0xFFFFFFFF;	// 			- fully incompatible
							
	/** Checks if two SubTiles can be mixed in single anomaly */
	public abstract int typeBits();
	
	public final void writeToNBT(NBTTagCompound cmp) {
		cmp.setInteger(TAG_TICKS, ticks);
		writeCustomNBT(cmp);
	}
	
	public void writeCustomNBT(NBTTagCompound cmp) {}

	public final void readFromNBT(NBTTagCompound cmp) {
		ticks = cmp.getInteger(TAG_TICKS);
		readCustomNBT(cmp);
	}
	
	public void readCustomNBT(NBTTagCompound cmp) {}

	public static SubTileEntity forName(String name) {
		try {
			return AlfheimAPI.getAnomaly(name).newInstance();
		} catch (Exception e) {
			e.printStackTrace();
			return null;
		}
	}
	
	public World worldObj() {
		return superTile.getWorldObj();
	}
	
	public int x() {
		return superTile.xCoord;
	}
	
	public int x(double x) {
		return x() + MathHelper.floor_double(x);
	}
	
	public int y() {
		return superTile.yCoord;
	}
	
	public int y(double y) {
		return y() + MathHelper.floor_double(y);
	}
	
	public int z() {
		return superTile.zCoord;
	}
	
	public int z(double z) {
		return z() + MathHelper.floor_double(z);
	}
	
	
	// ################################ UTILS ################################
	
	public EntityLivingBase findNearestVulnerableEntity(double radius) {
        List<EntityLivingBase> list = allAround(EntityLivingBase.class, radius);
        EntityLivingBase entity1 = null;
        double d0 = Double.MAX_VALUE;

        for (EntityLivingBase entity2 : list) {
            if (entity2 != null) {
            	if (entity2.isEntityInvulnerable()) continue;
            	if (entity2 instanceof EntityPlayer && ((EntityPlayer) entity2).capabilities.disableDamage) continue;
            	
                double d1 = Vector3.entityTileDistance(entity2, superTile);

                if (d1 <= d0) {
                    entity1 = entity2;
                    d0 = d1;
                }
            }
        }

        return entity1;
    }
	
	public EntityLivingBase findNearestEntity(double radius) {
        List<EntityLivingBase> list = allAround(EntityLivingBase.class, radius);
        EntityLivingBase entity1 = null;
        double d0 = Double.MAX_VALUE;

        for (EntityLivingBase entity2 : list) {
            if (entity2 != null) {
            	
                double d1 = Vector3.entityTileDistance(entity2, superTile);

                if (d1 <= d0) {
                    entity1 = entity2;
                    d0 = d1;
                }
            }
        }

        return entity1;
    }
	
	public <E> List<E> allAround(Class<E> clazz, double radius) {
		return worldObj().getEntitiesWithinAABB(clazz, AxisAlignedBB.getBoundingBox(x(), y(), z(), x(1), y(1), z(1)).expand(radius, radius, radius));
	}
	
	public List allAroundRaw(Class clazz, double radius) {
		return worldObj().getEntitiesWithinAABB(clazz, AxisAlignedBB.getBoundingBox(x(), y(), z(), x(1), y(1), z(1)).expand(radius, radius, radius));
	}
}

File: src/main/java/alfheim/api/block/tile/SubTileEntity.kt
package alfheim.api.block.tile

import alexsocol.asjlib.math.Vector3
import alfheim.api.AlfheimAPI
import alfheim.api.lib.LibResourceLocations
import net.minecraft.client.Minecraft
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.*
import net.minecraft.world.World

import java.util.*

// Used for anomalies - TileAnomaly
abstract class SubTileEntity {
	
	val rand = Random()
	var superTile: TileEntity? = null
	var ticks: Int = 0
	var worldGen = false
	
	abstract val targets: List<Any>
	
	abstract val rarity: EnumAnomalityRarity
	
	val frames: Int
		get() = 32
	
	open val strip: Int
		get() = 0
	
	val color: Int
		get() = 0xFFFFFF
	
	enum class EnumAnomalityRarity {
		COMMON, RARE, EPIC
	}
	
	/** optional update method for particles or other stuff  */
	protected open fun update() {}
	
	fun updateEntity(l: MutableList<Any?>?) {
		update()
		
		try {
			if (l == null || l.isEmpty()) return
			while (l.contains(null)) l.remove(null)
			if (l.isEmpty()) return
			
			for (target in l) performEffect(target!!)
		} finally {
			ticks++
		}
	}
	
	abstract fun performEffect(target: Any)
	
	/** Checks if two SubTiles can be mixed in single anomaly  */
	abstract fun typeBits(): Int
	
	open fun onActivated(stack: ItemStack?, player: EntityPlayer, world: World, x: Int, y: Int, z: Int) = false
	
	fun writeToNBT(cmp: NBTTagCompound) {
		cmp.setInteger(TAG_TICKS, ticks)
		writeCustomNBT(cmp)
	}
	
	open fun writeCustomNBT(cmp: NBTTagCompound) {}
	
	fun readFromNBT(cmp: NBTTagCompound) {
		ticks = cmp.getInteger(TAG_TICKS)
		readCustomNBT(cmp)
	}
	
	open fun readCustomNBT(cmp: NBTTagCompound) {}
	
	// ################################ SUPERTILE ################################
	
	val worldObj
	get() = superTile!!.worldObj!!
	
	fun x() = superTile!!.xCoord
	
	fun x(x: Double) = x() + MathHelper.floor_double(x)
	
	fun y() = superTile!!.yCoord
	
	fun y(y: Double) = y() + MathHelper.floor_double(y)
	
	fun z() = superTile!!.zCoord
	
	fun z(z: Double) = z() + MathHelper.floor_double(z)
	
	// ################################ UTILS ################################
	
	fun findNearestVulnerableEntity(radius: Double): EntityLivingBase? {
		val list = allAround(EntityLivingBase::class.java, radius)
		var entity1: EntityLivingBase? = null
		var d0 = java.lang.Double.MAX_VALUE
		
		for (entity2 in list) {
			if (entity2.isEntityInvulnerable) continue
			if (entity2 is EntityPlayer && entity2.capabilities.disableDamage) continue
			
			val d1 = Vector3.entityTileDistance(entity2, superTile!!)
			
			if (d1 <= d0) {
				entity1 = entity2
				d0 = d1
			}
		}
		
		return entity1
	}
	
	fun findNearestEntity(radius: Double): EntityLivingBase? {
		val list = allAround(EntityLivingBase::class.java, radius)
		var entity1: EntityLivingBase? = null
		var d0 = java.lang.Double.MAX_VALUE
		
		for (entity2 in list) {
			
			val d1 = Vector3.entityTileDistance(entity2, superTile!!)
			
			if (d1 <= d0) {
				entity1 = entity2
				d0 = d1
			}
		}
		return entity1
	}
	
	fun <E> allAround(clazz: Class<E>, radius: Double) =
		worldObj.getEntitiesWithinAABB(clazz, AxisAlignedBB.getBoundingBox(x().toDouble(), y().toDouble(), z().toDouble(), x(1.0).toDouble(), y(1.0).toDouble(), z(1.0).toDouble()).expand(radius, radius, radius)) as MutableList<E>
	
	fun allAroundRaw(clazz: Class<*>, radius: Double) =
		worldObj.getEntitiesWithinAABB(clazz, AxisAlignedBB.getBoundingBox(x().toDouble(), y().toDouble(), z().toDouble(), x(1.0).toDouble(), y(1.0).toDouble(), z(1.0).toDouble()).expand(radius, radius, radius)) as MutableList<Any>
	
	fun inWG() = worldGen
	
	// ################################ RENDER ################################
	
	fun bindTexture() {
		Minecraft.getMinecraft().renderEngine.bindTexture(LibResourceLocations.anomalies)
	}
	
	companion object {
		
		const val TAG_TICKS = "ticks"
		val EMPTY_LIST = ArrayList<Any>(0)
		
		const val NONE = 0
		//	0b00000 - fully compatible, do not use this unless you know what you are doing
		const val MOTION = 1
		//	0b00001 - motion manipulation	- gravity
		const val HEALTH = 2
		//	0b00010 - health manipulation	- damaging
		const val MANA = 4
		//	0b00100 - mana manipulation		- drain mana
		const val TIME = 8
		//	0b01000 - ticks manipulation	- time speedup
		const val SPACE = 17
		//	0b10001 - space manipulation	- teleportation		- also incompatible with motion
		const val ALL = -0x1    // 			- fully incompatible
		
		fun forName(name: String): SubTileEntity? {
			return try {
				AlfheimAPI.getAnomaly(name).newInstance()
			} catch (e: Exception) {
				e.printStackTrace()
				null
			}
		}
	}
}
